ci/ray_ci/pipeline/gap_filling_scheduler.py

```diff
@@ -0,0 +1,50 @@
+from datetime import datetime, timedelta
+from typing import List, Dict, Optional, Any
+
+from pybuildkite.buildkite import Buildkite
+
+from release.ray_release.aws import get_secret_token
+from release.ray_release.configs.global_config import AWS_SECRET_BUILDKITE
+
+
+class GapFillingScheduler:
+    """
+    This buildkite pipeline scheduler is responsible for scheduling gap filling builds
+    when the latest build is failing.
+    """
+
+    buildkite = None
+
+    def __init__(
+        self,
+        buildkite_organization: str,
+        buildkite_pipeline: str,
+    ):
+        self.buildkite_organization = buildkite_organization
+        self.buildkite_pipeline = buildkite_pipeline
+        GapFillingScheduler._init_buildkite()
+
+    @classmethod
+    def _init_buildkite(cls):
+        if not cls.buildkite:
+            cls.buildkite = Buildkite()
+            cls.buildkite.set_access_token(get_secret_token(AWS_SECRET_BUILDKITE))
+
+    def _get_latest_revision_for_build_state(self, build_state: str) -> Optional[str]:
```

Comment:
1. latest "commit" rather than "revision"?
2. what if there is a manual build on a commit from the past on the pipeline? like for recreating a wheel or re-testing something? I think to get the recent history of the branch and cross reference with it to get the correct logical order of the commits, rather than relying buildkite's build number.

Reply:
good point; the logic is a bit involved so let me put a TODO to make the logic more precise in subsequent PR, to avoid bloating this PR

Reply:
I worry that this logic is just wrong for these manually triggered corner cases, and can lead to a lot of unnecessary bisect runs when being used.

to allow you moving forward.. maybe for this PR, rather than picking one, just return the entire list of recent history?

Reply:
yes, i won't run this in production, just want to break it down for easier of review; i can also see how much additional logic is added, if not to much i can just add it here